keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,626 @@
+- Headless Kubernetes services: CoreDNS will resolve these services to either an IPv4 entry (A record), an IPv6 entry (AAAA record), or both, depending on the service's endpointFamily configuration (see [Configuration of Endpoint IP Family in Service Definitions](#configuration-of-endpoint-ip-family-in-service-definitions)).
+
+### Ingress Controller Operation
+
+The [Kubernetes ingress feature](https://kubernetes.io/docs/concepts/services-networking/ingress/) relies on the use of an ingress controller. The two "reference" ingress controllers that are considered here are the [GCE ingress controller](https://github.com/kubernetes/ingress-gce/blob/master/README.md#glbc) and the [NGINX ingress controller](https://github.com/kubernetes/ingress-nginx/blob/master/README.md#nginx-ingress-controller).
+
+#### GCE Ingress Controller: Out-of-Scope, Testing Deferred For Now
+It is not clear whether the [GCE ingress controller](https://github.com/kubernetes/ingress-gce/blob/master/README.md#glbc) supports external, dual-stack access. Testing of dual-stack access to Kubernetes services via a GCE ingress controller is considered out-of-scope until after the initial implementation of dual-stack support for Kubernetes.
+
+#### NGINX Ingress Controller - Dual-Stack Support for Bare Metal Clusters
+The [NGINX ingress controller](https://github.com/kubernetes/ingress-nginx/blob/master/README.md#nginx-ingress-controller) should provide dual-stack external access to Kubernetes services that are hosted on baremetal clusters, with little or no changes.
+
+- Dual-stack external access to NGINX ingress controllers is not supported with GCE/GKE or AWS cloud platforms.
+- NGINX ingress controller needs to be run on a pod with dual-stack external access.
+- On the load balancer (internal) side of the NGINX ingress controller, the controller will load balance to backend service pods on a per dual-stack-endpoint basis, rather than load balancing on a per-address basis. For example, if a given backend pod has both an IPv4 and an IPv6 address, the ingress controller will treat the IPv4 and IPv6 address endpoints as a single load-balance target. Support of dual-stack endpoints may require upstream changes to the NGINX ingress controller.
+- Ingress access can cross IP families. For example, an incoming L7 request that is received via IPv4 can be load balanced to an IPv6 endpoint address in the cluster, and vice versa. 
+
+### Load Balancer Operation
+
+As noted above, External load balancers that rely on Kubernetes services for load balancing functionality will only work with the IP family that matches the IP family of the cluster's service CIDR.
+
+#### Type ClusterIP
+
+The ClusterIP service type will be single stack, so for this case there will be no changes to the current load balancer config. The user has the option to create two load balancer IP resources, one for IPv6 and the other for IPv4, and associate both with the same application instances.
+
+#### Type NodePort
```

Comment:
Curious - how important is this requirement?  What if NodePorts were single-stack, at least initially?  Woudl it make this more tractable?

Reply:
I think that support of dual-stack NodePorts is an important requirement, specifically for providing dual-stack access from outside of the cluster. For example, if you look at the way that the NGINX ingress controller can be configured on bare metal (see [Bare-metal considerations](https://kubernetes.github.io/ingress-nginx/deploy/baremetal/), there are several installation options (MetalLB, over NodePort, via host network, self-provisioned edge, over externalIPs, etc.), but all of these options except the host network option require dual-stack nodePorts or externalIPs.

Supporting dual-stack nodePorts will require some code refactoring (or change in workflow) in kube-proxy (i.e. should be able to set up NodePort in iptables without requiring simultaneous configuration of service IPs), but I don't think that will be major work.

We could consider phasing this in later, but IMO dual-stack Kubernetes support isn't useful without dual-stack nodePorts and externalIPs.

Reply:
Sadly, that is what I thought you'd say, which means we NEED dual-stack endpoints, right?

Reply:
Yes, I believe so.